### DISCUSSION TITLE
Porting a Win9x video driver to OpenWatcom V2

### DISCUSSION THREAD

**Discussion by Danaozhong** [open-watcom/open-watcom-v2#1011] [Q&A]
Hello everyone!
I am currently working on modifying a Win9x video driver to be buildable with OpenWatcom V2. I want to be able to build it under Windows 10.
The video driver comes with a 16bit mini driver component. Microsoft intended this part to be compiled with MSVC 152. You can find a sample of the (nmake) makefile include for the 16bit mini driver here: https://github.com/KunYi/W2K_DDK_SAMPLES/blob/ee9eb5e572d1abdddc99cb9215fd21557f206dae/win_me/display/mini/mini16.mk 
This link is a sample display driver provided by MS, and my video driver is basically using a modified example. The 16 bit code will get compiled to a `*.drv` library.

I have modified my `mini16.mk` to use the OpenWatcom2 `wcc` compiler, and use the OpenWatcom2 include directory and library directory, but I am still facing problems.
I have the following questions:
1) Can I link object files generated by the MASM assembler with object files generated by OW2?
2) Can I use the MSVC152 linker to link MASM and OW2 object files, or shall I use the OW2 linker? 
3) MSVC 152 provides a C library `ldllcew.lib`, which contains the C standard library in large memory model, to be used when building as a library. I believe the equivalent library from OW2 should `clibl.lib`, so I changed:
```
DRVLIBS = $(DRVLIBS) \
          kernel.lib \
          dibeng.lib \
          ldllcew.lib
```
to
```
DRVLIBS = $(DRVLIBS) \
          kernel.lib \
          dibeng.lib \
          clibl.lib
```

However it seems `clibl.lib` doesn't work when building libraries, as I receive a linker error:
`C:\TOOL\OpenWatcom20\lib286\dos\clibl.lib(cmain086.c) : error L2029: 'main_' : unresolved external`
May I ask which `clib` I should use if I want to build my code as a library?

The full linker log is attached.

I am very new to driver development and OpenWatcom 2. Thanks for your help, and for keeing the development of OW active!

[linker_log.txt](https://github.com/open-watcom/open-watcom-v2/files/10199116/linker_log.txt)

**Comment by jmalak** [comment]
I think main mistake is that you link it as executable but it should be linked as DLL.
Anyway it is not as simple as it looks like.
Generaly you can use MASM with OW, no problem.
But there can be issue with interfacing to C code because Open Watcom C compiler uses different calling convention then MS and  different symbol names mangling and different C run-time startup code.

How you compile and link it by Open Watcom?

**Reply by Danaozhong** [reply]
Thank you! I noticed that the calling convention is different, but saw that you can pass a compiler flag to change the calling convention. This also resulted in almost all linker errors being resolved, so I think it is the right way. Just for completeness, these are the compiler parameters passed to the MSVC 152 compiler:
`CFLAGS = -GD -Alfw  -Z7i -W3 -c -Zp -Gf3s -nologo -Fc`
For OW2, I am using the following compiler flags:
`CFLAGS = -bt=windows -W=1 -Ze -3 -j -ecd -ml`

Which should mean (if I understood correctly):
```
# -bt = build target=16 bit Windows
# -j change char default from unsigned to signed 
# -ml large memory model - needed to be compatible to the MASM generated files
# -ecd __stcall calling convention - needed to be compatible to the MASM generated files
# -Ze enable extensions
```

**Reply by Danaozhong** [reply]
Regarding linking, I am currently still using the MSVC 152 linker (which surprisingly works under Win10). I would like to switch to the OW linker, but as you mentioned, it requires some migration effort.
For completeness, the MSVC 152 linker gets invoked with the following parameters (which I so far have no clue what they mean):
`L16FLAGS = /CO/NOE/NOD/NOPACKCODE/NOPACKFUNCTIONS/LI/MAP/AL:16 $(L16DEF)`

**Reply by Danaozhong** [reply]
May I ask how I can link it as a DLL? Is there a flag that I have to pass to the linker/compiler?

**Reply by jmalak** [reply]
If you link by OW linker then specify **`system windows_dll`** to link 16-bit DLL (NE format).
To compile files for DLL you should add **`-bd`** compiler option which include DLL startup code reference to properly link DLL.

**Comment by jmalak** [comment]
Anyway simplest solution is to compile and link it with MS tools.
If you wan't to use Open Watcom tools then some porting is necessary, it required some knowledge about differences between MS nad OW tools and C run-time libraries.
It looks like that code is by design compiled and linked as 16-bit Windows DLL.
OW by default compile and link for 16-bit DOS executable that you need to handle it to be correct 16-bit Windows target.
Also I don't know what dibeng.lib library is, it is not OW library and can be trouble with it.
You can transform MS link command to OW wlink by ms2wlink tool, it also convert used .def files to appropriate wlink directives.

**Reply by Danaozhong** [reply]
Just to answer your question: `dibeng.lib` and `KERNEL.LIB` are both part of the Win9x DDK. Thank you for the suggestion on how to convert the linker command! I will give it a try.

**Reply by jmalak** [reply]
If any problem with linking then add OW linker map file which is generated by **`option map`** OW linker option

**Comment by Danaozhong** [comment]
Thank you very much for your help! It helped me a lot to make some progress!

- I migrated to the OW2 linker, with the help of `ms2wlink`. The linker file is attached below.
[watcom_linker_file2.lnk.txt](https://github.com/open-watcom/open-watcom-v2/files/10200390/watcom_linker_file2.lnk.txt)

- I added the `-bd` flag to indicate that I am building a DLL.
- I had to specify `format windows dll initglobal` instead of `system windows_dll` (that did result in the following error:
 `Warning! W1107: file ..\watcom_linker_file2.lnk.txt: line(2): undefined system name: windows_dll`)


I am still facing the issue when linking `clibl.lib` that it requires a symbol `_main`.

Linker output:
```
H:\Projekte\2021\3dfx_Treiber\Git\Win9x_Driver\DD16\debug>C:\TOOL\OpenWatcom20\binnt\wlink.exe @..\watcom_linker_file2.lnk.txt
Open Watcom Linker Version 2.0 beta Nov 21 2022 00:54:00 (32-bit)
Copyright (c) 2002-2022 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
Warning! W1107: file ..\watcom_linker_file2.lnk.txt: line(2): undefined system name: windows_dll
loading object files
searching libraries
Warning! W1008: cannot open emu87.lib : No such file or directory
Warning! W1008: cannot open math87l.lib : No such file or directory
Warning! W1008: cannot open clibl.lib : No such file or directory
Error! E2030: file C:\TOOL\OpenWatcom20\lib286\win\clibl.lib(wstart): multiple starting addresses found, already defined in diblink.obj(..\diblink.asm)
Warning! W1008: cannot open commdlg.lib : No such file or directory
Error! E2028: __DLLstart_ is an undefined reference
Error! E2028: main_ is an undefined reference
creating map file
creating a Windows dynamic link library
file C:\TOOL\OpenWatcom20\lib286\win\clibl.lib(wmain.c): undefined symbol main_
```

The map file is below:
[display.map.txt](https://github.com/open-watcom/open-watcom-v2/files/10200376/display.map.txt)

I am bit surprised that even the library built using MSVC 152 does not have a `DLLstart` symbol. I attached the `display.drv` that was built under Win9x using MSVC152 for comparison. I am not sure what kind of library `*.drv` actually is. Perhaps they are different to normal DLLs?

[display_compiled_with_msvc152_for_comparison.zip](https://github.com/open-watcom/open-watcom-v2/files/10200381/display_compiled_with_msvc152_for_comparison.zip)

**Reply by jmalak** [reply]
The message "Warning! W1107: file ..\watcom_linker_file2.lnk.txt: line(2): undefined system name: windows_dll)" indicate that you don't have setup WATCOM or PATH environment variables to appropriate OW subdirectory.
WATCOM should point to root of OW installation and PATH should point to binary subdirectory for host system.
For windows it is %WATCOM%\binnt64 (64-bit version) or %WATCOM%\binnt (32-bit version)
you can use **`format windows dll`** instead of **`system windows_dll`** in your case.

**Comment by jmalak** [comment]
I am sorry, I don't do any Windows 16/32-bit driver design that my note will be only general.
It looks like diblink.asm reference driver startup code somehow.
Please could you give here map file generated by MS linker.

**Reply by jmalak** [reply]
I did a small code investigation and it looks like 16-bit DLL has no standard startup code.
It looks like it is 16-bit code for using by 32-bit driver part with chunks to call 16-bit code in 16-bit DLL.
it will be necessary somehow excluded OW standard startup code from linking to not include CRTL startup code.
Therefore you should use **`-zl -zls`** OW C compiler options instead of **`-bd`** option.

**Reply by Danaozhong** [reply]
Thank you! Here is a linker file (for comparison) for the same library, built and linked with MSVC 152.
[msvc_152_display.map.txt](https://github.com/open-watcom/open-watcom-v2/files/10201249/msvc_152_display.map.txt)

I have tried setting options `-zl -zls`, and removing `-bd`:
1) If I link `clibl.lib`, I still get the linker errors:

```
Error! E2030: file C:\TOOL\OpenWatcom20\lib286\win\clibl.lib(wstart): multiple starting addresses found, already defined in diblink.obj(..\diblink.asm)
Warning! W1008: cannot open emu87.lib : No such file or directory
Warning! W1008: cannot open math87l.lib : No such file or directory
Warning! W1008: cannot open clibl.lib : No such file or directory
Warning! W1008: cannot open commdlg.lib : No such file or directory
Error! E2028: main_ is an undefined reference
creating map file
creating a Windows dynamic link library
file C:\TOOL\OpenWatcom20\lib286\win\clibl.lib(wmain.c): undefined symbol main_
```

2) Without `clibl.lib`, I get lots of linker errors:
```
Error! E2028: __STK is an undefined reference
Error! E2028: __U4M is an undefined reference
Error! E2028: __I4M is an undefined reference
Error! E2028: __I4D is an undefined reference
Error! E2028: __U4D is an undefined reference
Error! E2028: atoi_ is an undefined reference
Error! E2028: strcat_ is an undefined reference
Error! E2028: strcpy_ is an undefined reference
Error! E2028: memset_ is an undefined reference
Error! E2028: _fmemcpy_ is an undefined reference
Error! E2028: atol_ is an undefined reference
Error! E2028: _itoa_ is an undefined reference
Error! E2028: strlen_ is an undefined reference
[...]
```
I will have a look into the C library to understand why `_main` is needed, and if there is an easy way to remove it from linking.

**Reply by jmalak** [reply]
You must recompile all object files because there are old references from previous compilations.
It is reason to use -zl -zls which disable generation of reference to default libraries and default symbols, but you must recompile all.
That if it refers to any library or main symbols then it is wrong or missing -zl -zls options.
-zl           remove default library information
-zls          remove automatically inserted symbols

**Reply by Danaozhong** [reply]
Thank you - I have done a fresh build, but the driver code is actually using parts of the standard library (`strlen`, `atoi`, `strcpy`, `memset`, and others).

**Reply by jmalak** [reply]
I think it is correct, but it should not link CRTL startup code.

**Comment by Danaozhong** [comment]
I found out that this part of the display driver handles the DLL initialization, and also does some magic to with an `__acrtused` symbol. This seems to be very Microsoft MSVC152-specific code to handle library initialization and remove the initialization code of the MS C standard library.
[DIBLINK.ASM.txt](https://github.com/open-watcom/open-watcom-v2/files/10201310/DIBLINK.ASM.txt)
I get the feeling that this is now very Microsoft-driver-development specific, so let me dive a bit more into that. If possible, please keep this thread open for a bit - once I found a solution, I would like to post it here.

**Reply by jmalak** [reply]
No problem, you can use it for next problem.
Maybe you will need to add some OW default symbols to diblink to not load OW C run-time startup code or unused modules at all.
